### PR TITLE
Initial RISC-V support

### PR DESCRIPTION
**Pull request by athei**
This makes ink! compile on RISC-V. There are two things which are just filled with interim values:

- Location and size of the heap
- Function ids and ABI of the host functions

Both of these things are not final and will be replaced once we have written the specification of our VM. Nonetheless we should still merge the code once this PR is in a good state: This allows us to iterate with the new RISC-V VM and prevent RISC-V from breaking in the future.

I will also need to make changes to `cargo contract`. As of right now I am compiling manually. Once a new cargo contract version is released I can add the CI jobs to this PR. Those will replace our `no-std` stop gap CI jobs.

Notable changes required for support:

- RISC-V does not support `cdylib`. Instead, we build contracts as bin target (to be added by cargo contract). We need to use a single entry point in case of RISC-V. This is also why every contract needs to add `no_main` in addition to `no_std` when building for RISC-V (wasm doesn't care). Wasm can have multiple entry points even when build as bin. So we can do that unconditionally.
- Panic handler needs support for RISC-V
- Memory allocator needs support for RISC-V
- oom handler needs support for RISC-V. I opted for removing the handler and raising the Rust version. A default handler is supplied starting with version 1.68. Hence we no longer need the unstable feature.

### CODE REVIEW

**Comment by athei** [comment]
Apparently doc tests are not run if a library doesn't include the `rlib`target. By adding it to all integration tests (by falling back to default) the multisig doc tests were suddenly run. I fixed them up.

Also needed to update the dyling toolchain as we no require 1.68.

**Comment by athei** [comment]
Ready for review. Please get this in as-is because it touches a lot of files and I expect merge conflicts. I will do a follow up PR where I will use `cargo contract` in the CI.

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -39,10 +39,12 @@ variables:
   # this fix not yet in stable: https://github.com/rust-lang/rust-clippy/issues/8895.
   # Remove the following line again as soon as `clippy` on stable succeeds again.
   CLIPPY_ALLOWED:                  "clippy::extra_unused_lifetimes"
-  # This is a target that does not have a standard library in contrast to Wasm. We compile against
-  # this target to make sure that we don't pull in `std` by accident (through dependencies).
-  # See https://doc.rust-lang.org/nightly/rustc/platform-support.html#tier-2.
-  NO_STD_TARGET:                   "aarch64-unknown-none"
+  # We plan to fully support RISC-V as a bytecode for contracts soon.
+  # RISC-V does not have a standard library in contrast to Wasm. Compiloing against
```
Comment (ascjones) [inline .gitlab-ci.yml:43]:
```suggestion
  # RISC-V does not have a standard library in contrast to Wasm. Compiling against
```

**Review by ascjones — APPROVED** [review]

File: crates/env/src/engine/on_chain/ext/riscv32.rs [hunk 2]
```diff
@@ -0,0 +1,410 @@
...
+        $(
+
+            pub fn $name(output: &mut &mut [u8]) {
+                let mut output_len = output.len() as u32;
+                {
+					(
+						Ptr32Mut::from_slice(output),
+						Ptr32Mut::from_ref(&mut output_len),
+					).using_encoded(|in_data| sys::call(FUNC_ID, Ptr32::from_slice(in_data)));
+                }
```
[the remark this answers is not in the document]
Comment (athei) [reply crates/env/src/engine/on_chain/ext/riscv32.rs:271]: What can I do about it if this is how `rustfmt` wants it?

[the remark this answers is not in the document]
Comment (athei) [reply crates/env/src/engine/on_chain/ext/riscv32.rs:271]: Didn't realize this was inside a macro. I think indentation is correct. We just have a superfluous block. Removed.

**Comment by athei** [comment]
The code for riscv uses allocation to provide a buffer to encode the arguments for the host into. This will be changed to a static buffer in a follow up. It requires a bigger refactor.

**Pull request merged**